### PR TITLE
Add 32-bit and 80-bit floating point support to LLVMVisitor

### LINKED ISSUE #1605: Lambdify with other precision?
I'm wondering if it's possible to provide versions of Lambdify where backend='llvm' also takes precision as an argument? I'm asking since I'm interested in computing a reference solution using "long double". Also, being able to calculate using 32-bit precision would be interesting since I have some CUDA code which I run on consumer graphics cards (batched LU decomposition), not having to convert back and forth would probably shave off some computational time.

### PR DESCRIPTION
**Pull request by isuruf**
Fixes #1605

### CODE REVIEW

**Comment by bjodah** [comment]
This is looking good. I thought I had a decent grip on floating point representations, but I was surprised that:
```
#include <stdio.h>
int main() {
    printf("%lu\n", sizeof(long double));
}
```
actually prints 16 on my Intel CPU (I would have guessed to get 10 on x86 FPU, and perhaps 16 on PowerPC), so a long double actually occupies 16 bytes on x86 (but only has 80 bits in its representation). A bit of searching suggests this is due to Linux x86 ABI:
https://stackoverflow.com/a/53506801/790973
(giving better alignment). Not sure what the story is on windows...

**Comment by bjodah** [comment]
What about `LLVMVisitor::set_double` and `mp_get_d`. (or maybe it becomes a slippery slope?)

File: symengine/llvm_double.cpp [hunk 1]
```diff
@@ -345,43 +343,89 @@ void LLVMDoubleVisitor::call(double *outs, const double *inps)
...
+#ifndef HAVE_SYMENGINE_MPFR
+    throw NotImplementedError("Cannot convert to long double without MPFR");
+#else
+    RCP<const Basic> m = evalf(x, 128, EvalfDomain::Real);
+    result_ = llvm::ConstantFP::get(get_float_type(&mod->getContext()),
+                                    m->__str__());
+#endif
+}
+
+void LLVMLongDoubleVisitor::visit(const Integer &x)
```
[the remark this answers is not in the document]
Comment (bjodah) [reply symengine/llvm_double.cpp]: Nice, missed that.

**Comment by bjodah** [comment]
regarding `set_double`: yeah I missed that first time around, but noticed after you pointed out `mp_get_d` not being called. Looks real good.
Oh, didn't know that about MSVC, interesting. [Looks like](https://stackoverflow.com/a/30096769/790973) mingw-w64 does support it?

File: symengine/llvm_double.cpp [hunk 2]
```diff
@@ -898,37 +984,57 @@ void LLVMDoubleVisitor::loads(const std::string &s)
...
+
+llvm::Type *LLVMFloatVisitor::get_float_type(llvm::LLVMContext *context)
+{
+    return llvm::Type::getFloatTy(*context);
+}
+
+#if defined(SYMENGINE_HAVE_LLVM_LONG_DOUBLE)
+llvm::Type *LLVMLongDoubleVisitor::get_float_type(llvm::LLVMContext *context)
+{
+    return llvm::Type::getX86_FP80Ty(*context);
```
Comment (bjodah) [inline symengine/llvm_double.cpp]: We don't have access to Power9 hardware to test their FP128 type right?

**Comment by isuruf** [comment]
On i686, `long double` is 12 bytes. I've changed the check now.

**Comment by isuruf** [comment]
> Looks like mingw-w64 does support it?

Then, this PR should work. You'll probably need an LLVM with a mingw default target.

File: symengine/llvm_double.cpp (hunk 2 above)
Comment (bjodah) [reply symengine/llvm_double.cpp]: So we're only supporting long double on x86 for now? Just checking, sounds painful to fix CI for something that's unofficially supported anyhow.

**Comment by isuruf** [comment]
> Do we want to rename the files from llvm_double to e.g. llvm_visitor?

Is that necessary? I'd prefer to keep the name as is to avoid changing the name in downstream projects.

**Comment by bjodah** [comment]
No, keeping llvm_double is fine, just asking. +1 to merge! (really nice PR)

**Comment by isuruf** [comment]
Thanks for the review

**Pull request merged**